src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1900,85 +1900,102 @@ impl Cage {
+                    Stream(obj) => &mut obj.flags,
+                    File(obj) => &mut obj.flags,
+                    //not clear why running F_SETFL on Socket type requires special treatment
+                    Socket(ref mut sockfdobj) => {
+                        if cmd == F_SETFL && arg >= 0 {
+                            let sock_tmp = sockfdobj.handle.clone();
+                            let mut sockhandle = sock_tmp.write();
+
+                            if let Some(ins) = &mut sockhandle.innersocket {
+                                let fcntlret;
+                                if arg & O_NONBLOCK == O_NONBLOCK {
+                                    //set for non-blocking I/O
+                                    fcntlret = ins.set_nonblocking();
+                                } else {
+                                    //clear non-blocking I/O
+                                    fcntlret = ins.set_blocking();
+                                }
+                                if fcntlret < 0 {
+                                    match Errno::from_discriminant(interface::get_errno()) {
+                                        Ok(i) => {
+                                            return syscall_error(
+                                                i,
+                                                "fcntl",
+                                                "The libc call to fcntl failed!",
+                                            );
+                                        }
+                                        Err(()) => panic!("Unknown errno value from fcntl returned!"),
+                                    };
+                                }
                             }
                         }
-                    }
 
-                    &mut sockfdobj.flags
-                }
-            };
+                        &mut sockfdobj.flags
+                    }
+                };
 
-            //matching the tuple
-            match (cmd, arg) {
-                //because the arg parameter is not used in certain commands, it can be anything (..)
-                (F_GETFD, ..) => *flags & O_CLOEXEC,
-                // set the flags but make sure that the flags are valid
-                (F_SETFD, arg) if arg >= 0 => {
-                    if arg & O_CLOEXEC != 0 {
-                        *flags |= O_CLOEXEC;
-                    } else {
-                        *flags &= !O_CLOEXEC;
+                //matching the tuple
+                match (cmd, arg) {
+                    //because the arg parameter is not used in certain commands, it can be anything (..)
+                    //currently, O_CLOEXEC is the only defined file descriptor flag, thus only this flag is
+                    //masked when using F_GETFD or F_SETFD
+                    (F_GETFD, ..) => *flags & O_CLOEXEC,
+                    // set the flags but make sure that the flags are valid
```

Comment:
I'm not sure how this achieves your goal.  Can you say more?

Reply:
F_GETFD command should return file descriptor flags only, which means that  the returned value should not include access mode flags and file status flags, and because O_CLOEXEC is the only implemented file descriptor as of now, we can get it by bit-wise and'ing.